### PR TITLE
Introducing `vsmtp-protocol`

### PR DESCRIPTION
**Pull request by Mathieu-Lala**
# Context

This PR is a refactoring of the SMTP receiver part of the application.

By merging, we did : 
* eliminate most of the technical depth
* separate the protocol (`vsmtp-protocol`) and the policies implementation (`vsmtp-rule-engine`)
* enhance performance (need more tests)

# Summary of changes

## Added

* `vsmtp-protocol` containing a generic `ESMTPSA` implementation, with a `trait RecevierHandler` for simpler integration of the policies treatment

## Changed

* use `str::eq_ignore_ascii_case` instead of `str::to_lowercase == str::to_lowercase`
* the `MailParser::parse` take a `Stream<Item = std::io::Result<Vec<u8>>>`
* the `MessageBody`'s headers terminate by `\r\n` now

### CODE REVIEW

File: src/vsmtp/vsmtp-rule-engine/src/rule_state.rs [hunk 1]
```diff
@@ -15,69 +15,23 @@
...
 
 impl RuleState {
-    /// create a new rule state with connection data.
-    #[must_use]
-    pub fn with_connection(rule_engine: &RuleEngine, conn: Connect) -> Self {
-        let mut state = rule_engine.spawn();
-
-        // all rule are skipped until the designated rule
-        // in case of a delegation result.
-        #[cfg(feature = "delegation")]
```
Comment (Mathieu-Lala) [inline src/vsmtp/vsmtp-rule-engine/src/rule_state.rs:48]: Call this routine

File: src/vsmtp/vsmtp-rule-engine/src/rule_state.rs [hunk 2]
```diff
@@ -15,69 +15,23 @@
...
-use vsmtp_common::status::Status;
 use vsmtp_mail_parser::MessageBody;
 use vsmtp_plugins::rhai;
 
 /// a state container that bridges rhai's & rust contexts.
+#[derive(Debug)]
 pub struct RuleState {
     /// A lightweight engine for evaluation.
-    pub(super) engine: rhai::Engine,
+    pub(super) engine: std::sync::Arc<rhai::Engine>,
```
Comment (ltabis) [inline src/vsmtp/vsmtp-rule-engine/src/rule_state.rs:25]: Remove Arc

File: src/vsmtp/vsmtp-server/src/server.rs [hunk 3]
```diff
@@ -253,29 +251,46 @@ impl Server {
...
             queue_manager,
-            &mut MailHandler {
-                working_sender,
-                delivery_sender,
-            },
-        )
-        .await?;
+        );
+        let smtp_receiver = vsmtp_protocol::Receiver::<_, ValidationVSL, _, _>::new(
+            tcp_stream,
```
Comment (Mathieu-Lala) [inline src/vsmtp/vsmtp-server/src/server.rs:282]: Construct the `AcceptArgs` here

File: src/vsmtp/vsmtp-rule-engine/src/dsl/directives/mod.rs [hunk 4]
```diff
@@ -46,6 +46,14 @@ pub enum Directive {
     },
 }
 
+impl std::fmt::Debug for Directive {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct(self.as_ref())
+            .field("name", &self.name())
```
Comment (ltabis) [inline src/vsmtp/vsmtp-rule-engine/src/dsl/directives/mod.rs:52]: Could also add the pointer address.

**Review by ltabis — APPROVED** [review]

**Pull request merged**